larreco/TrackFinder/PMAlgTrackMaker_module.cc

```diff
@@ -0,0 +1,1368 @@
+//
+// Progress:
+//    May-June 2015:   track finding and validation, growing tracks by iterative merging of matching
+//                     clusters, no attempts to build multi-track structures, however cosmic tracking
+//                     works fine as they are sets of independent tracks
+//    June-July 2015:  merging track parts within a single tpc and stitching tracks across tpc's
+//    August 2015:     optimization of track-vertex structures (so 3D vertices are also produced)
+//    November 2015:   use track-shower splitting at 2D level, then tag low-E EM cascades in 3D
+//                     note: the splitter is not finished and not as good as we want it
+//    January 2016:    output of track-vertex finding as a tree of PFParticles, refined vertexing
+//                     code, put vertex at front of each track, flip whole track structures to
+//                     selected, direction (beam/down/dQdx), use any pattern reco stored in
+//                     PFParticles as a source of associated clusters
+//    Mar-Apr 2016:    kinks finding, EM shower direction reconstructed for PFPaarticles tagged as
+//                     electrons
+//
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art/Framework/Services/Optional/TFileService.h"
+#include "art/Framework/Services/Optional/TFileDirectory.h"
+#include "art/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// LArSoft includes
+#include "larcore/Geometry/Geometry.h"
+#include "larcore/Geometry/TPCGeo.h"
+#include "larcore/Geometry/PlaneGeo.h"
+#include "larcore/Geometry/WireGeo.h"
+#include "lardata/RecoBase/Hit.h"
+#include "lardata/RecoBase/Cluster.h"
+#include "lardata/RecoBase/PFParticle.h"
+#include "lardata/RecoBase/Track.h"
+#include "lardata/RecoBase/TrackHitMeta.h"
+#include "lardata/RecoBase/Vertex.h"
+#include "lardata/RecoBase/SpacePoint.h"
+#include "lardata/AnalysisBase/T0.h" 
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "lardata/Utilities/AssociationUtil.h"
+
+#include "larreco/RecoAlg/ProjectionMatchingAlg.h"
+#include "larreco/RecoAlg/PMAlgVertexing.h"
+#include "larreco/RecoAlg/PMAlg/Utilities.h"
+#include "larreco/RecoAlg/PMAlg/PmaTrkCandidate.h"
+
+#include "TTree.h"
+#include "TMath.h"
+
+#include <memory>
+
+namespace trkf {
+
+typedef std::map< size_t, std::vector<double> > dedx_map;
+typedef std::map< unsigned int, std::vector< art::Ptr<recob::Hit> > > view_hitmap;
+typedef std::map< unsigned int, view_hitmap > tpc_view_hitmap;
+typedef std::map< unsigned int, tpc_view_hitmap > cryo_tpc_view_hitmap;
```

Comment:
A followup to my note above.  If you need to support the possibility that the hit Ptr's point to objects in more than one data product, I see two classes of solutions

1. If the data products are all std::vectors, then the transient representation can be a a bare pointer.  We would need to provide the appropriate helper to reconstitute Ptrs.
2. Another solution is to define a "transient Ptr" class that has the persistent info plus a bare pointer; it would not do any validity check so its as fast to use as a bare pointer. Again we would have to add helpers to reconstitute Ptrs.